[Messenger][DX] Uses custom method names for handlers

sroze opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/pull/26685#issuecomment-383888657
| License       | MIT
| Doc PR        | ø

This has been discussed mostly in the [`MessageHandlerInterface` pull-request](https://github.com/symfony/symfony/pull/26685). For consistency reasons and convenience, this PR adds the ability to configure the method to be used on handlers:
```php
use Symfony\Component\Messenger\Handler\MessageSubscriberInterface;
use Symfony\Component\Messenger\Handler\MessageSubscriberConfiguration;

class CreateNumberMessageHandler implements MessageSubscriberInterface
{
    /**
     * {@inheritdoc}
     */
    public static function getHandledMessages(): array
    {
        return [
            CreateNumber::class => ['createNumber', 10],
            AnotherMessage::class => 'anotherMethod',
        ];
    }

    public function createNumber(CreateNumber $command)
    {
        // ...
    }
}
```

sroze committed:
[Messenger][DX] Uses custom method names for handlers

ogizanagi reviewed (commented):
`MessageSubscriberInterface::getHandledMessages` docblock misses an update

chalasr commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -60,27 +61,49 @@ public function process(ContainerBuilder $container)
 
-                foreach ($handles as $messageClass) {
+                foreach ($handles as $messageClass => $method) {
+                    if (is_int($messageClass)) {
```
let's add a `\` in front of `is_*` calls to be consistent with the existing one above

chalasr commented on src/Symfony/Component/Messenger/Handler/MethodOnObjectHandler.php:
```diff
@@ -0,0 +1,38 @@
+    public function __construct($object, string $method)
+    {
+        if (!is_object($object)) {
+            throw new \InvalidArgumentException(sprintf('Expected an object as argument but got %s', gettype($object)));
```
`\is_object`, `\gettype` (no, I'm not a bot :))

sroze commented on src/Symfony/Component/Messenger/Handler/MethodOnObjectHandler.php:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Used to transform an object and a method to a callable handler.
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
```
Good point, added 👍

Tobion commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -60,27 +61,55 @@ public function process(ContainerBuilder $container)
+                    if ('__invoke' !== $method) {
+                        $wrapperDefinition = (new Definition(MethodOnObjectHandler::class))->setArguments(array(new Reference($serviceId), $method));
+
+                        $definitions[$serviceId = hash('sha1', $messageClass.':'.$messagePriority.':'.$serviceId)] = $wrapperDefinition;
```
It looks like it does not work for 

```
public static function getHandledMessages(): array
    {
        return [
            CreateNumber::class => 'method',
            CreateNumber::class => 'anotherMethod',
        ];
    }
```

nicolas-grekas replied:
also, please use ContainerBuilder::hash instead of sha1

stof replied:
@Tobion Your array does not work. It has a single item in it as PHP does not allow duplicate keys.

however, the hash should indeed include the method name, as you can have multiple methods.

and the final service id should not be only the hash. It should have a prefix before it to make it easier to identify services in error message (`messenger.method_object_handler.<hash>` could make sense)

sroze replied:
Good point, method should be in the hash. Updated 👍

Tobion commented:
From my experience the EventSubscriber return value to configure callbacks has a really bad UX. It's not clear what array structure can be returned without reading the doc. Repeating that in the `MessageSubscriberInterface` seems bad to me.

How about using a value object like

```php
    public static function getHandledMessages(): array
    {
        return [
            new MessageCallback(CreateNumber::class, 'createNumber', 10),
            new MessageCallback(AnotherMessage::class, 'anotherMethod'),
        ];
    }
```

```php
class MessageCallback
{
    public function __construct(string $messageClass, string $method = '__invoke', int $priority = 0)
    {
    }
}
```

This way the code is self-explaining and you don't have to deal with mixed returns like array of array or array of string.

sroze commented:
I like it!

Tobion commented:
Learning from mistakes and improving them seems more important to me. Also the EventSubscriber with it's array structure is probably not gonna make it like this into the EventDispatcher PSR that got revived. But something like my propal is really flexible and clear. Just make MessageCallback an interface with getters and voila.

nicolas-grekas commented:
public static function getHandledMessages(): HandledMessagesDescription
`return (new HandledMessagesDescription())->someFluentApi()...`?

but this looks like a lot of code infrastructure, which means this works only with a compiled container, because when not compiled the overhead of building this configuration can become significant, especially compared to the array approach, isn't it?

Tobion commented:
I don't see where the overhead is when using a value object or a fluent api? This is just setters and getters.

sroze commented:
@nicolas-grekas I prefer the array approach as it's much simpler (for everybody - to use and to create) and this isn't a complex building mechanism that requires such thing I believe.

stof commented:
@sroze the array of EventSubscriberInteface is very simple when you have to register a single listener. But when you need to register 2 listeners for the same event, I admit that I get it wrong half of the time (doing `array('foo', 'bar')` instead of `array(array('foo'), array('bar'))`. I even went as far as writing a unit test taking all my subscribers and validating the structure being returned.

nicolas-grekas commented:
Would it make sense to use the static method as a configurator?
```php
static function configureHandledMessages(HandledMessagesDescriptor $config)
{
    $config->...
}
```

BUT on the other end, any objects we put here tightly couples the interface to the descriptor.
It could be a blocker when taking interoperability into account: by forcing an object wrapper, we forbid any custom extension of the return value. An array doesn't have this drawback.
(just thinking loud :) )

Tobion commented:
One more idea that closes the circle in my eyes:

Remove https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Messenger/Handler/MessageHandlerInterface.php and instead offer a trait/abstract class that implements 

```php
public static function getHandledMessages(): iterable
{
    return [
        new ReflectedMethodHandler(self::class, '__invoke'),
    ];
}
```

nicolas-grekas commented:
I think this is building infrastructure on top of conventions. We should design first and implement after IMHO.
Another issue with using objects is that it will make it much harder for IDEs to use the return value to provide auto-completion.
On a DX perspective, I really much prefer injecting the descriptor than have a "new" in the body of the method. The reason is that "new" forces to know *what* type to create while passing by argument is just perfect for providing autocompletion.

Disclaimer: the arguments I'm giving here are done while taking the existing static method descriptions we already have. @Tobion uses the word "mistake" above, I'm really not sure this was a mistake at all. And more importantly, I want to be sure whatever new system we use is actually better.

nicolas-grekas commented:
As @Tobion spotted, using an injected argument enforces a single style of configuration (the DX is bound to the style provided by`HandledMessagesDescriptor` in my example above). While this can be nice for autocompletion, I don't think we can make a universal enough style (that could eg fit a FIG PSR - which is not a goal here, but still nice for guiding the thought).

I'm also not really convinced by returning objects, for similar reasons: a locally created object forces also a style, or if we make it an interface, the interface could only provide a way to retrieve the state of the configuration, without providing any contractual guarantee (unlike the injected argument btw).

For this reasons, I now think this method should return an array.
I also think we can provide a builder for this array if we want to, but this builder should be an (optional) implementation detail used in some implementations of `getHandledMessages()`.

kbond commented:
> For this reasons, I now think this method should return an array.

Why not iterable?

nicolas-grekas commented:
Or iterable, just low-level data structure :)

sroze commented:
I've introduced a `MessageSubscriberConfiguration` object that can be used to specify the priority and/or method name. To reduce the complexity, I've removed the ability to have priority or method name from the tags.

Tobion commented:
Three things:

1. There should be an interface for `MessageSubscriberConfiguration` which only requires the getters. Most people will obviously use the `MessageSubscriberConfiguration` value object implementation but thats an implementation detail that should not be part of the `MessageSubscriberInterface`
2. Returning a string instead of MessageSubscriberConfiguration should not be supported anymore. No mixed return values please.
3.  I think the name `getHandledMessages` is not optimal. But I can't think of a better alternative currently.

Tobion commented on src/Symfony/Component/Messenger/Handler/MessageSubscriberInterface.php:
```diff
@@ -23,18 +23,28 @@
+     * If not configured otherwise, the `__invoke` method of the handler will be called as usual with
+     * the message to handle.
      *
      * @return array
```
`MessageSubscriberConfigurationInterface[]` or a better name

sroze commented:
> There should be an interface for MessageSubscriberConfiguration [...] value object

I appreciate the potentiality of such interface at some point in the future but as you mentioned, this is a value object and I don’t believe we need an interface for that (we can add it later in case it’s useful, without any BC break).

sroze commented:
> Returning a string instead of MessageSubscriberConfiguration should not be supported anymore. No mixed return values please.

I’m actually not sure about this one. On a developer perspective, the usual would be to simply define the message class; therefore making it easier (than having to instanciate this value object) seems logical. On the other hand, the “simple option” is to just use the type hint on `_invoke`, that’s probably enough (and a good practice to promote anyway).

@symfony/deciders by using 👍 and 👎 on this comment, do you think it’s a good idea to drop the ability to use the class name as a string in the message subscribers’ static method?

xabbuh commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -21,8 +21,10 @@
+use Symfony\Component\Messenger\Handler\MessageSubscriberConfiguration;
 use Symfony\Component\Messenger\Handler\MessageSubscriberInterface;
 use Symfony\Component\Messenger\TraceableMessageBus;
+use Symfony\Component\Messenger\Handler\MethodOnObjectHandler;
```
should be before the import of the `TraceableMessageBus` class

nicolas-grekas commented:
In https://github.com/symfony/symfony/pull/27034#issuecomment-385229190 I gave my reasoning explaining why I think returning object is not a good idea. Instead, I suggest to use a builder:
return (new FooBuilder())->add(...)->add()->toArray().
or
return [HandledMessageDescription::build(...), etc], with build() returning an array.

Basically, the getter on MessageSubscriberConfiguration should not be exposed as they have no meaning (they are just techinal requirements to acces the state) and thus they should not be part of the subscriber interfaces.

Tobion commented:
> if we make it an interface, the interface could only provide a way to retrieve the state of the configuration, without providing any contractual guarantee

I don't understand what you mean? Could you clarify?
So you propose to return some random array structure again similar to EventSubscriber?

Tobion commented:
> But since we need a technical means to access the state of the VO on the outside, we must add getters on it. And this is a leak in the abstraction.

I don't see the difference between getters and an array with keys. Both leak information in the same way. I'm not sure what you are trying to achieve here.

nicolas-grekas commented:
@sroze if you want to move forward here, I suggest to isolate things that are unrelated to the return format of the static method in another PR if that makes sense.

sroze commented:
@nicolas-grekas i.e. revert to the array-notation, right?

sroze commented:
Alright, that's fair that there are different point of views and it's a tricky question, therefore let's make that in two steps: I reverted the introduction of the VO and kept the array EventDispatcher-like syntax (while changing the `getHandledMessages()` signature to return a `iterable` rather than an `array`).

sroze commented:
Shall we go ahead with this one?

ogizanagi reviewed (approved):
Yes, makes sense to me as is. Let's move forward 👍

nicolas-grekas commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -72,27 +73,55 @@ public function process(ContainerBuilder $container)
+                    if ('__invoke' !== $method) {
+                        $wrapperDefinition = (new Definition(MethodOnObjectHandler::class))->setArguments(array(new Reference($serviceId), $method));
+
+                        $definitions[$serviceId = 'messenger.method_on_object_handler.'.ContainerBuilder::hash($messageClass.':'.$messagePriority.':'.$serviceId.':'.$method)] = $wrapperDefinition;
```
Missing dot at start of service is.

nicolas-grekas replied:
But can't we pass array($def, $method) instead of using this handler?

sroze replied:
Not the way it's done now because we simply wire the handlers (i.e. `callables`). That way is also simpler :)

sroze commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -23,6 +23,7 @@
 use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
 use Symfony\Component\Messenger\Handler\MessageSubscriberInterface;
 use Symfony\Component\Messenger\TraceableMessageBus;
+use Symfony\Component\Messenger\Handler\MethodOnObjectHandler;
```
updated 👍

nicolas-grekas commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -72,27 +73,55 @@ public function process(ContainerBuilder $container)
+                    }
+
+                    if ('__invoke' !== $method) {
+                        $wrapperDefinition = (new Definition(MethodOnObjectHandler::class))->setArguments(array(new Reference($serviceId), $method));
```
`$wrapperDefinition = (new Definition('callable'))->setArguments(array(new Reference($serviceId), $method))->setFactiry('Closure::fromCallable');`
cheers :)

sroze replied:
:dance:

fabpot commented on src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:
```diff
@@ -72,27 +72,55 @@ public function process(ContainerBuilder $container)
+                        $method = '__invoke';
+                    }
+
                     if (\is_array($messageClass)) {
```
could be `else if` here, right?

sroze replied:
Correct 👍

sroze replied:
Changed.

sroze replied:
@fabpot actually can't because we are setting `$messagePriority` as well.

sroze commented:
Status: Needs review

sroze commented:
@fabpot could you update your review please?
